From 0e9bd6014d94b637839ab0e9d72ba0e220e154f5 Mon Sep 17 00:00:00 2001
From: Andrew Reynolds <andrew.j.reynolds@gmail.com>
Date: Thu, 9 Jan 2025 10:56:50 -0600
Subject: [PATCH 284/312] Minor fixes for Eunoia signature and output (#11503)

Also updates the ethos version to the current one, which has some
improved error messages.
---
 contrib/get-ethos-checker           | 2 +-
 proofs/eo/cpc/programs/Utils.eo     | 8 +++++---
 proofs/eo/cpc/theories/Builtin.eo   | 2 +-
 proofs/eo/cpc/theories/Datatypes.eo | 2 +-
 src/proof/alf/alf_printer.cpp       | 1 +
 5 files changed, 9 insertions(+), 6 deletions(-)

diff --git a/contrib/get-ethos-checker b/contrib/get-ethos-checker
index 8c8c15e09..b99a6425a 100755
--- a/contrib/get-ethos-checker
+++ b/contrib/get-ethos-checker
@@ -26,7 +26,7 @@ EO_DIR="$BASE_DIR/ethos-checker"
 mkdir -p $EO_DIR
 
 # download and unpack ethos
-ETHOS_VERSION="ffd697b162e556883ed26958b052a5b8ad7bdc77"
+ETHOS_VERSION="7b9ea06ab3c5e39b66366396cd216e3f8637c64e"
 download "https://github.com/cvc5/ethos/archive/$ETHOS_VERSION.tar.gz" $BASE_DIR/tmp/ethos.tgz
 tar --strip 1 -xzf $BASE_DIR/tmp/ethos.tgz -C $EO_DIR
 
diff --git a/proofs/eo/cpc/programs/Utils.eo b/proofs/eo/cpc/programs/Utils.eo
index d4690bd3e..6783fc6da 100644
--- a/proofs/eo/cpc/programs/Utils.eo
+++ b/proofs/eo/cpc/programs/Utils.eo
@@ -89,9 +89,11 @@
 ; define: $tail
 ; args:
 ; - x S: The term to inspect.
-; return: the tail of x, where x is expected to be a function application.
-(define $tail ((S Type :implicit) (x S))
-  (eo::match ((T Type) (U Type) (S Type) (f (-> T U S)) (x1 T) (x2 T :list))
+; return: >
+;   The tail of x, where x is expected to be an application of an n-ary
+;   function marked :right-assoc-nil.
+(define $tail ((U Type :implicit) (x U))
+  (eo::match ((T Type) (U Type) (f (-> T U U)) (x1 T) (x2 U :list))
     x
     (((f x1 x2) x2)))
 )
diff --git a/proofs/eo/cpc/theories/Builtin.eo b/proofs/eo/cpc/theories/Builtin.eo
index b125ac185..0a005d890 100644
--- a/proofs/eo/cpc/theories/Builtin.eo
+++ b/proofs/eo/cpc/theories/Builtin.eo
@@ -55,7 +55,7 @@
 
 ; generic variables
 ; NOTE: does not check that U is a numeral
-(declare-const const (-> (! Type :var U :implicit) U (! Type :var T) T))
+(declare-const const (-> (! Type :var U :implicit) (! U :opaque) (! Type :opaque :var T) T))
 
 ; The purification skolem.
 (declare-const @purify (-> (! Type :var A :implicit) (! A :opaque) A))
diff --git a/proofs/eo/cpc/theories/Datatypes.eo b/proofs/eo/cpc/theories/Datatypes.eo
index 8b396d30c..fd4f41f41 100644
--- a/proofs/eo/cpc/theories/Datatypes.eo
+++ b/proofs/eo/cpc/theories/Datatypes.eo
@@ -60,4 +60,4 @@
 (declare-const nullable.lift (-> (! Type :var F :implicit) F ($get_type_nullable_lift F)))
 
 ; shared selector
-(declare-const @shared_selector (-> (! Type :var D) (! Type :var T) Int D T))
+(declare-const @shared_selector (-> (! Type :opaque :var D) (! Type :opaque :var T) (! Int :opaque) D T))
diff --git a/src/proof/alf/alf_printer.cpp b/src/proof/alf/alf_printer.cpp
index fd7df1882..05c99d349 100644
--- a/src/proof/alf/alf_printer.cpp
+++ b/src/proof/alf/alf_printer.cpp
@@ -730,6 +730,7 @@ void AlfPrinter::print(AlfPrintChannelOut& aout,
         smt::PrintBenchmark pb(nodeManager(), &alfp, false, &d_tproc);
         std::stringstream outDecl;
         std::stringstream outDef;
+        options::ioutils::applyPrintArithLitToken(outDef, true);
         pb.printDeclarationsFrom(outDecl, outDef, definitions, assertions);
         out << outDecl.str();
         // [2] print the definitions
-- 
2.47.1

